quinn/src/connection.rs

```diff
@@ -324,22 +324,40 @@ impl Connection {
     /// Streams are cheap and instantaneous to open unless blocked by flow control. As a
     /// consequence, the peer won't be notified that a stream has been opened until the stream is
     /// actually used.
-    pub fn open_uni(&self) -> OpenUni {
-        OpenUni {
-            conn: self.0.clone(),
-            state: broadcast::State::default(),
-        }
+    pub async fn open_uni(&self) -> Result<SendStream, ConnectionError> {
+        let (id, is_0rtt) = self.open(Dir::Uni).await?;
+        Ok(SendStream::new(self.0.clone(), id, is_0rtt))
     }
 
     /// Initiate a new outgoing bidirectional stream.
     ///
     /// Streams are cheap and instantaneous to open unless blocked by flow control. As a
     /// consequence, the peer won't be notified that a stream has been opened until the stream is
     /// actually used.
-    pub fn open_bi(&self) -> OpenBi {
-        OpenBi {
-            conn: self.0.clone(),
-            state: broadcast::State::default(),
+    pub async fn open_bi(&self) -> Result<(SendStream, RecvStream), ConnectionError> {
+        let (id, is_0rtt) = self.open(Dir::Bi).await?;
+        Ok((
+            SendStream::new(self.0.clone(), id, is_0rtt),
+            RecvStream::new(self.0.clone(), id, is_0rtt),
+        ))
+    }
+
+    async fn open(&self, dir: Dir) -> Result<(StreamId, bool), ConnectionError> {
+        loop {
+            let opening;
+            {
+                let mut conn = self.0.lock("open");
+                if let Some(ref e) = conn.error {
+                    return Err(e.clone());
+                }
+                if let Some(id) = conn.inner.streams().open(dir) {
+                    let is_0rtt = conn.inner.side().is_client() && conn.inner.is_handshaking();
+                    return Ok((id, is_0rtt));
+                }
+                opening = conn.stream_opening[dir as usize].clone();
+                opening.notified() // Must be called, but not awaited, while the connection lock is held
```

Comment:
After all the refactoring the core of this change is now much clearer but I don't have a good intuition for what we are looping against here and why we need the awkward binding declaration and clone. Maybe add just a sprinkling of comments to communicate the intent here?

Reply:
Added detailed discussion.